File: app/clicDB/src/java/edu/xtec/web/clic/userlib/GetUserInfo.java
/**
 * GetUserInfo.java
 * 
 * Checks if the token provided by the user is valid and, when true, returns
 * information about the user's zone status and projects currently published on
 * it.
 *
 **/
package edu.xtec.web.clic.userlib;

import edu.xtec.util.db.ConnectionBean;
import edu.xtec.web.clic.Context;
import edu.xtec.web.clic.Utilities;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.URL;
import java.sql.PreparedStatement;
import java.text.DateFormat;
import java.util.Date;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONStringer;

/**
 *
 * @author fbusquet
 */
public class GetUserInfo extends HttpServlet {

  // Default disk quota is 50 MB
  public static final int DEFAULT_QUOTA = 52428800;

  // Parameter name for OAuth token
  public static final String ID_TOKEN = "id_token";
  // Endpoint to validate OAuth tokens sent by Google
  public static final String CHECK_GOOGLE_TOKEN = "https://www.googleapis.com/oauth2/v3/tokeninfo?id_token=";
  // Root directory of user's zone in the local filesystem
  private static File ROOT_BASE = null;

  // Information fields about the current user
  public String email;
  public String userId;
  public String fullUserName;
  public String avatar;
  public Date expires;
  public String expDate;
  public long quota = DEFAULT_QUOTA;

  // UserSpace object assignated to the current user
  public UserSpace userSpace;

  // Main servlet method. Reacts only to "POST" queries.
  protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
    // Response will be always a JSON
    response.setContentType("application/json");
    response.setCharacterEncoding("UTF-8");
    PrintWriter writer = response.getWriter();
    try {
      loadUserData(request);
      writer.write(getJsonUserInfo().toString());
      logMsg("INFO", "User logged in");
    } catch (Exception ex) {
      // Hand-made JSON, used to return the error message
      writer.print("{\"status\":\"error\",\"error\":\"" + JSONStringer.getString(ex.getMessage()) + "\"}");
      logMsg("ERROR", "Login incorrect: " + ex.getMessage());
    } finally {
      writer.flush();
    }
  }

  // Loads the data associated with the current user
  protected void loadUserData(HttpServletRequest request) throws Exception {
    String token = Utilities.getParam(request, ID_TOKEN, null);
    if (token != null) {
      // Validate token
      URL verifyURL = new URL(CHECK_GOOGLE_TOKEN + token);
      JSONObject json = Utilities.readJSON(verifyURL.openStream());

      // Check for valid email
      email = json.getString("email");
      if (email == null) {
        throw new Exception("Usuari desconegut");
      }
      // "hd" field is used only with GSuite users. Contains the domain name.
      String hd = json.optString("hd", "");

      // Read global settings
      File settingsFile = new File(Context.getStaticFileBase(), Context.cntx.getProperty("userLibCfg", "users/settings.json"));
      JSONObject settings = Utilities.readJSON(new FileInputStream(settingsFile));
      // Get the current generic disk quota
      quota = settings.optLong("quota", quota);
      // Check if user is allowed to login, and load specific user settings
      boolean validUser = "xtec.cat".equals(hd);
      JSONArray usr = settings.getJSONArray("users");
      for (int i = 0; i < usr.length(); i++) {
        JSONObject usrObj = usr.getJSONObject(i);
        String id = usrObj.optString("id", "");
        if (email.equals(id)) {
          validUser = true;
          quota = usrObj.optLong("quota", quota);
          break;
        }
      }
      if (!validUser) {
        // Unathorized used
        throw new Exception("Usuari no autoritzat: " + email);
      }

      // Get user ID from email
      userId = getPlainId(email, "xtec.cat");
      // Initialize the UserSpace object
      userSpace = new UserSpace(userId, new File(getRootBase(), userId));
      userSpace.readProjects();

      // Get additional data
      fullUserName = json.getString("name");
      avatar = json.optString("picture", null);
      expires = new Date(json.getLong("exp") * 1000);

      // Create a session object and save current user's data on it
      setSessionAttributes(request.getSession(true));

    } else {
      // No token presented
      // Try to read settings from a stored session, if any
      getSessionAttributes(request.getSession(false));
    }
  }

  // Encapsulates current user data in a JSON expression, ready to be sent
  public JSONObject getJsonUserInfo() throws Exception {
    JSONObject json = new JSONObject();
    if (email == null) {
      json.put("status", "error");
      json.put("error", "invalid user");
    } else {
      json.put("email", email);
      json.put("id", userId);
      json.put("quota", quota);
      json.put("fullUserName", fullUserName);
      json.putOpt("avatar", avatar);
      json.put("expires", DateFormat.getDateTimeInstance().format(expires));
      json.put("status", "validated");
      json.put("projects", userSpace.getProjectsJSON());
      json.put("currentSize", userSpace.currentSize);
    }
    return json;
  }

  // Save current user data in current session object
  protected void setSessionAttributes(HttpSession session) throws Exception {
    if (session != null) {
      // Retain sessions only for 30'
      session.setMaxInactiveInterval(1800);

      // Save user settings in the HttpSession object
      session.setAttribute("email", email);
      session.setAttribute("id", userId);
      session.setAttribute("quota", new Long(quota));
      session.setAttribute("fullUserName", fullUserName);
      session.setAttribute("avatar", avatar == null ? "" : avatar);
      session.setAttribute("expires", expires);
      session.setAttribute("userSpace", userSpace);
    }
  }

  // Retrieve user data from current session object, if any
  protected void getSessionAttributes(HttpSession session) throws Exception {
    if (session != null) {
      email = (String) session.getAttribute("email");
      userId = (String) session.getAttribute("id");
      quota = ((Long) session.getAttribute("quota")).longValue();
      fullUserName = (String) session.getAttribute("fullUserName");
      avatar = (String) session.getAttribute("avatar");
      expires = (Date) session.getAttribute("expires");
      userSpace = (UserSpace) session.getAttribute("userSpace");
    }
  }

  // Miscellaneous functions
  // Gets a simplified user ID from its e-mail address
  // Only plain letters, digits and dots are allowed
  public static String getPlainId(String email, String hd) {
    email = email.trim().toLowerCase();
    int p = email.lastIndexOf("@" + hd);
    if (p > 0) {
      email = email.substring(0, p);
    } else {
      email = email.replace('@', '.');
    }
    char[] ch = email.toCharArray();
    for (int i = 0; i < ch.length; i++) {
      char c = ch[i];
      if (c < '0' || (c > '9' && c < 'a') || c > 'z') {
        ch[i] = '.';
      }
    }
    return new String(ch);
  }

  // Initialize the ROOT_BASE variable, reading it from "Context" if needed
  public static File getRootBase() throws IOException {
    if (ROOT_BASE == null) {
      ROOT_BASE = new File(Context.getStaticFileBase(), Context.cntx.getProperty("userLibRoot", "users"));
      if (!ROOT_BASE.canWrite()) // Root directory must be writable
      {
        throw new IOException("Invalid root base!");
      }
    }
    return ROOT_BASE;
  }

  /* ---------------------------------------

Log table created with:
------------------------------------------  
CREATE TABLE  "LOG_USERACTIONS" 
   (	"LOG_DATE" DATE NOT NULL ENABLE, 
	"LOG_TYPE" CHAR(5) NOT NULL ENABLE, 
	"LOG_USER" VARCHAR2(100), 
	"LOG_MSG" VARCHAR2(1024)
   )
/

CREATE INDEX  "LOG_USERACTIONS_LOG_DATE_IDX" ON  "LOG_USERACTIONS" ("LOG_DATE")
/

CREATE INDEX  "LOG_USERACTIONS_LOG_TYPE_IDX" ON  "LOG_USERACTIONS" ("LOG_TYPE")
/
------------------------------------------ */
  // Appends a message to the log table
  protected void logMsg(String type, String msg) {
    ConnectionBean con = null;
    PreparedStatement stmt = null;
    try {
      con = Context.cntx.getDB().getConnectionBean();
      stmt = con.getPreparedStatement("INSERT INTO CLIC.LOG_USERACTIONS (LOG_DATE, LOG_TYPE, LOG_USER, LOG_MSG) VALUES (?,?,?,?)");
      stmt.setDate(1, new java.sql.Date(System.currentTimeMillis()));
      stmt.setString(2, type);
      stmt.setString(3, userId == null ? "unknown" : userId);
      stmt.setString(4, msg);
      stmt.execute();
    } catch (Exception ex) {
      System.err.println("Unable to write logs to DB due to: " + ex.getMessage());
    } finally {
      if (con != null) {
        try {
          con.closeStatement(stmt);
          Context.cntx.getDB().freeConnectionBean(con);
        } catch (Exception ex) {
          System.err.println("Unable to close DB connection due to: " + ex.getMessage());
        }
      }
    }
  }

}

File: app/db/userlib/getUserInfo.php
<?php
/**
 * File: getUserInfo.php
 * 
 * Linted with PHP_CodeSniffer (http://pear.php.net/package/PHP_CodeSniffer/)
 * against PEAR standards (https://pear.php.net/manual/en/standards.sample.php)
 * with phpcbf
 * 
 * PHP Version 7
 * 
 * @category Service
 * @package  UserLib
 * @author   Francesc Busquets <francesc@gmail.com>
 * @license  https://www.tldrlegal.com/l/eupl-1.1 EUPL-1.1
 * @link     https://github.com/projectestac/zonaclic
 */


// *****************************************************************
// ** WARNING: 'allow_url_fopen' must be set to 'On' in 'php.ini' **
// *****************************************************************

require_once '../config.php';
require_once '../log.php';
require_once 'userSpace.php';

/**
 * Gets a simplified user ID from its e-mail address
 * Only plain letters, digits and dots are allowed
 * 
 * @param string $email Full user's e-mail
 * @param string $hd    The G Suite organization where the user is enrolled, if any
 * 
 * @return string
 */
function getPlainId($email, $hd)
{
    $r = strtolower(trim($email));
    $p = strrpos($r, '@'.$hd);
    $r = $p === false ? str_replace('@', '.', $r) : substr($r, 0, $p);

    $result = '';
    foreach (str_split($r) as $ch) {
        $result .= (($ch < '0' || ($ch > '9' && $ch < 'a') || $ch > 'z') ? '.' : $ch);
    }
    
    return $result;
}

/**
 * Gets the value associated to $key in the $array if set,
 * returning $default otherwise
 * 
 * @param array  $array   The array where to search for
 * @param string $key     Key to be searched
 * @param any    $default Default value to return if key not found
 * 
 * @return any
 */
function getAttr($array, $key, $default)
{
    return isset($array->{$key}) ? $array->{$key} : $default;
}

/**
 * Gets the content of a remote file, performing a network call
 * 
 * @param string $url The remote file to retrieve
 * 
 * @return string
 */
function getRemoteFile($url)
{
    $result = '';
    try {
        if (!USE_PROXY) {
            $result = file_get_contents($url);
        } else {
            $proxy = PROXY_HOST.':'.PROXY_PORT;
            $ch = curl_init();
            curl_setopt($ch, CURLOPT_URL, $url);
            curl_setopt($ch, CURLOPT_PROXY, $proxy);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            //curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
            //curl_setopt($ch, CURLOPT_USERAGENT,'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.13) Gecko/20080311 Firefox/2.0.0.13');
            $result = curl_exec($ch);
            curl_close($ch);
            if ($result === '' || $result === false) {
                throw new Exception('No s\'ha pogut verificar l\'usuari. Proxy error.');
            }
        }
    } catch (Exception $e) { 
        throw $e;
    }
    return $result;
}

/**
 * Main function
 */
if (isset($_POST[ID_TOKEN]) && $_POST[ID_TOKEN] !== '') {
    $errMsg = '';
    try {
        $result = (object)['status'=>'processing'];

        // Read settings file
        $settingsFileName = UserSpace::usersRoot().'/'.SETTINGS_FILE;
        $settings = json_decode(file_get_contents($settingsFileName), false);
    
        // Check token validity (Warning: external call to a Google API!)
        $raw = getRemoteFile(CHECK_GOOGLE_TOKEN.$_POST[ID_TOKEN]);
        $user = json_decode($raw, false);

        if (!isset($user->{'email'}) || $user->{'email'} === '') {
            // Invalid token!
            $result->status = 'error';
            $result->error = 'Validació incorrecta';
            $errmsg = ' No email in user data';
        } else {       
            // Check if user is valid and load email and quota
            $email = $user->email;
            $quota = getAttr($settings, 'userQuota', DEFAULT_QUOTA);
            $validUser = false;
       
            $hd = getAttr($user, 'hd', null);
            if ($hd === HD) {
                $validUser = true;
            }
       
            if (isset($settings->users)) {
                foreach ($settings->users as $usr) {
                    if ($usr->id === $email) {
                        $validUser = true;
                        $quota = getAttr($usr, 'quota', $quota);
                        break;
                    }
                }
            }

            if ($validUser !== true) {
                // Unathorized
                $result->status = 'error';
                $result->error = 'Usuari no autoritzat';
                $errMsg = 'User email: '.$email;
            } else if (!isset($user->exp) || $user->exp < (new DateTime())->getTimestamp()) {
                $result->status = 'error';
                $result->error = 'Credencial no vàlida';
                $errMsg = 'Expired session';
            } else {
                $result->email = $email;
                $result->quota = $quota;
                $result->id = getPlainId($email, HD);
                $result->fullUserName = $user->name;
                $result->avatar = getAttr($user, 'picture', '');
                $result->expires = date('M d, Y h:i:s A', $user->exp);
                $result->status = 'validated';

                $space = new UserSpace($result->id);
                $space->readProjects();
                $result->projects = $space->getProjectsData();
                $result->currentSize = $space->currentSize;

                // Set session data
                session_start();
                $_SESSION['userId'] = $result->id;
                $_SESSION['rootDir'] = $space->rootDir;                
                $_SESSION['quota'] = $result->quota;
                $_SESSION['currentSize'] = $result->currentSize;

                logMsg('LOGIN', 'user: '.$result->id.' ('.number_format($result->currentSize/MB, 2).'/'.number_format($result->quota/MB, 2).')');
            }
        }

        // Set response header and content
        header('Content-Type: application/json;charset=UTF-8');
        print json_encode($result);          

    } catch (Exception $e) {
        // Internal error
        http_response_code(500);
        print 'Internal error: '.$e->getMessage();
        logMsg('ERR-LOGIN', $e->getMessage().' '.$errMsg);
    }
} else {
    // Bad request
    http_response_code(400);
}
